library/constant_time_impl.h

```diff
@@ -31,7 +31,7 @@
  * Disable -Wredundant-decls so that gcc does not warn about this. This is re-enabled
  * at the bottom of this file.
  */
-#ifdef __GNUC__
+#if __GNUC__ > 4
```

Comment:
We should really check that the compiler is gcc - other compilers also define `__GNUC__`

```suggestion
#if defined(MBEDTLS_COMPILER_IS_GCC) && (__GNUC__ > 4)
```